File: finanze/application/use_cases/get_available_entities.py
from dataclasses import asdict
from datetime import datetime
from uuid import UUID

from application.ports.credentials_port import CredentialsPort
from application.ports.crypto_wallet_connection_port import CryptoWalletConnectionPort
from application.ports.entity_port import EntityPort
from application.ports.external_entity_port import ExternalEntityPort
from application.ports.last_fetches_port import LastFetchesPort
from application.ports.virtual_import_registry import VirtualImportRegistry
from dateutil.tz import tzlocal
from domain.available_sources import (
    AvailableSource,
    AvailableSources,
    FinancialEntityStatus,
)
from domain.entity import EntityOrigin, EntityType, Feature
from domain.external_entity import EXTERNAL_ENTITY_FEATURES, ExternalEntityStatus
from domain.native_entities import NATIVE_ENTITIES
from domain.use_cases.get_available_entities import GetAvailableEntities
from domain.virtual_fetch import VirtualDataImport


def get_last_fetches_for_virtual(
    virtual_imports: list[VirtualDataImport],
) -> dict[Feature, datetime]:
    last_fetch = {}
    for virtual_import in virtual_imports:
        last_fetch[virtual_import.feature] = virtual_import.date

    return last_fetch


class GetAvailableEntitiesImpl(GetAvailableEntities):
    LISTED_ENTITY_TYPES = [EntityType.FINANCIAL_INSTITUTION, EntityType.CRYPTO_WALLET]

    def __init__(
        self,
        entity_port: EntityPort,
        external_entity_port: ExternalEntityPort,
        credentials_port: CredentialsPort,
        crypto_wallet_connections_port: CryptoWalletConnectionPort,
        last_fetches_port: LastFetchesPort,
        virtual_import_registry: VirtualImportRegistry,
    ):
        self._entity_port = entity_port
        self._external_entity_port = external_entity_port
        self._credentials_port = credentials_port
        self._crypto_wallet_connections_port = crypto_wallet_connections_port
        self._last_fetches_port = last_fetches_port
        self._virtual_import_registry = virtual_import_registry

    def execute(self) -> AvailableSources:
        logged_entities = self._credentials_port.get_available_entities()
        logged_entity_ids = {e.entity_id: e.expiration for e in logged_entities}

        all_entities = self._entity_port.get_all()

        native_entities_by_id = {e.id: e for e in NATIVE_ENTITIES}

        last_virtual_imported_entities = self.get_last_virtual_imports_by_entity()

        entities = []
        for entity in all_entities:
            if entity.type not in self.LISTED_ENTITY_TYPES:
                continue
            native_entity = native_entities_by_id.get(entity.id)
            status = None
            wallets = None
            external_entity_id = None

            last_virtual_imported_data = last_virtual_imported_entities.get(entity.id)
            virtual_features = {}
            if last_virtual_imported_data:
                virtual_features = {
                    vi.feature: vi.date for vi in last_virtual_imported_data
                }

            dict_entity = asdict(native_entity or entity)

            if entity.origin == EntityOrigin.EXTERNALLY_PROVIDED:
                external_entity = self._external_entity_port.get_by_entity_id(entity.id)
                if not external_entity:
                    status = FinancialEntityStatus.DISCONNECTED
                    dict_entity["features"] = []
                else:
                    status = (
                        FinancialEntityStatus.CONNECTED
                        if external_entity.status == ExternalEntityStatus.LINKED
                        else FinancialEntityStatus.REQUIRES_LOGIN
                    )
                    external_entity_id = external_entity.id
                    dict_entity["features"] = EXTERNAL_ENTITY_FEATURES

            elif entity.type == EntityType.FINANCIAL_INSTITUTION:
                status = FinancialEntityStatus.DISCONNECTED

                if entity.origin != EntityOrigin.MANUAL:
                    if entity.id in logged_entity_ids:
                        status = FinancialEntityStatus.CONNECTED

                        expiration = logged_entity_ids.get(entity.id)
                        if expiration and expiration < datetime.now(tzlocal()):
                            status = FinancialEntityStatus.REQUIRES_LOGIN
                else:
                    if virtual_features:
                        status = FinancialEntityStatus.CONNECTED

            else:
                wallets = self._crypto_wallet_connections_port.get_by_entity_id(
                    entity.id
                )

            last_fetch = {}
            if entity.origin != EntityOrigin.MANUAL:
                if status != FinancialEntityStatus.DISCONNECTED:
                    last_fetch_records = self._last_fetches_port.get_by_entity_id(
                        entity.id
                    )
                    last_fetch = {r.feature: r.date for r in last_fetch_records}
            else:
                dict_entity["features"] = []

            entity_virtual_imports = last_virtual_imported_entities.get(entity.id)
            if entity_virtual_imports:
                virtual_last_fetch = get_last_fetches_for_virtual(
                    entity_virtual_imports
                )
                if entity.origin == EntityOrigin.MANUAL:
                    last_fetch = virtual_last_fetch
                    dict_entity["features"] = list(virtual_last_fetch.keys())

            entities.append(
                AvailableSource(
                    **dict_entity,
                    status=status,
                    connected=wallets,
                    last_fetch=last_fetch,
                    external_entity_id=external_entity_id,
                    virtual_features=virtual_features,
                )
            )

        return AvailableSources(entities=entities)

    def get_last_virtual_imports_by_entity(self) -> dict[UUID, list[VirtualDataImport]]:
        last_virtual_imports = self._virtual_import_registry.get_last_import_records()
        last_virtual_imported_entities = {}
        for virtual_import in last_virtual_imports:
            if virtual_import.entity_id not in last_virtual_imported_entities:
                last_virtual_imported_entities[virtual_import.entity_id] = []

            last_virtual_imported_entities[virtual_import.entity_id].append(
                virtual_import
            )

        return last_virtual_imported_entities

File: mobile/src/application/usecases/getAvailableEntities.ts
import {
  AvailableSource,
  AvailableSources,
  EntityOrigin,
  EntityType,
  ExternalEntityStatus,
  Feature,
  FinancialEntityStatus,
  ProductType,
  VirtualDataImport,
} from "@/domain"
import { NATIVE_ENTITIES } from "@/domain/nativeEntities"
import {
  CredentialsPort,
  CryptoWalletConnectionPort,
  EntityPort,
  ExternalEntityPort,
  LastFetchesPort,
  VirtualImportRegistry,
} from "../ports"
import { GetAvailableEntities } from "@/domain/usecases"

function getLastFetchesForVirtual(
  virtualImports: VirtualDataImport[],
): Partial<Record<Feature, string>> {
  const lastFetch: Partial<Record<Feature, string>> = {}
  for (const virtualImport of virtualImports) {
    if (!virtualImport.feature) continue
    lastFetch[virtualImport.feature] = virtualImport.date
  }
  return lastFetch
}

const EXTERNAL_ENTITY_FEATURES: Feature[] = [Feature.POSITION]

export class GetAvailableEntitiesImpl implements GetAvailableEntities {
  private static readonly LISTED_ENTITY_TYPES: EntityType[] = [
    EntityType.FINANCIAL_INSTITUTION,
    EntityType.CRYPTO_EXCHANGE,
    EntityType.CRYPTO_WALLET,
  ]

  private static readonly EXTERNAL_ENTITY_PRODUCTS: ProductType[] = [
    ProductType.ACCOUNT,
  ]
  private static readonly CRYPTO_WALLET_PRODUCTS: ProductType[] = [
    ProductType.CRYPTO,
  ]

  constructor(
    private entityPort: EntityPort,
    private externalEntityPort: ExternalEntityPort,
    private credentialsPort: CredentialsPort,
    private cryptoWalletConnectionsPort: CryptoWalletConnectionPort,
    private lastFetchesPort: LastFetchesPort,
    private virtualImportRegistry: VirtualImportRegistry,
  ) {}

  async execute(): Promise<AvailableSources> {
    const loggedEntities = await this.credentialsPort.getAvailableEntities()
    const loggedEntityIds = new Map<string, string | null>()
    for (const e of loggedEntities) {
      loggedEntityIds.set(e.entityId, e.expiration ?? null)
    }

    const allEntities = await this.entityPort.getAll()

    const nativeEntitiesById = new Map(
      NATIVE_ENTITIES.filter(e => !!e.id).map(e => [e.id as string, e]),
    )

    const lastVirtualImportedEntities =
      await this.getLastVirtualImportsByEntity()

    const entities: AvailableSource[] = []
    for (const entity of allEntities) {
      if (!entity.id) {
        continue
      }

      if (
        !entity.type ||
        !GetAvailableEntitiesImpl.LISTED_ENTITY_TYPES.includes(entity.type)
      ) {
        continue
      }

      const nativeEntity = nativeEntitiesById.get(entity.id)

      let status: FinancialEntityStatus | null = null
      let wallets: any = null
      let externalEntityId: string | null = null
      let products: ProductType[] | null | undefined = null

      const lastVirtualImportedData = lastVirtualImportedEntities.get(entity.id)
      let virtualFeatures: Partial<Record<Feature, string>> = {}
      if (lastVirtualImportedData && lastVirtualImportedData.length > 0) {
        virtualFeatures = getLastFetchesForVirtual(lastVirtualImportedData)
      }

      const dictEntity: any = { ...(nativeEntity ?? entity) }

      if (entity.origin === EntityOrigin.EXTERNALLY_PROVIDED) {
        products = GetAvailableEntitiesImpl.EXTERNAL_ENTITY_PRODUCTS
        const externalEntity = await this.externalEntityPort.getByEntityId(
          entity.id,
        )
        if (!externalEntity) {
          status = FinancialEntityStatus.DISCONNECTED
          dictEntity.features = []
        } else {
          status =
            externalEntity.status === ExternalEntityStatus.LINKED
              ? FinancialEntityStatus.CONNECTED
              : FinancialEntityStatus.REQUIRES_LOGIN
          externalEntityId = externalEntity.id
          dictEntity.features = EXTERNAL_ENTITY_FEATURES
        }
      } else if (
        entity.type === EntityType.FINANCIAL_INSTITUTION ||
        entity.type === EntityType.CRYPTO_EXCHANGE
      ) {
        status = FinancialEntityStatus.DISCONNECTED
        products = dictEntity.products

        if (entity.origin !== EntityOrigin.MANUAL) {
          if (loggedEntityIds.has(entity.id)) {
            status = FinancialEntityStatus.CONNECTED

            const expiration = loggedEntityIds.get(entity.id)
            if (expiration) {
              const expirationDate = new Date(expiration)
              if (expirationDate < new Date()) {
                status = FinancialEntityStatus.REQUIRES_LOGIN
              }
            }
          }
        } else {
          if (Object.keys(virtualFeatures).length > 0) {
            status = FinancialEntityStatus.CONNECTED
          }
        }
      } else {
        wallets = await this.cryptoWalletConnectionsPort.getByEntityId(
          entity.id,
        )
        products = GetAvailableEntitiesImpl.CRYPTO_WALLET_PRODUCTS
      }

      let lastFetch: Partial<Record<Feature, string>> = {}
      if (entity.origin !== EntityOrigin.MANUAL) {
        if (status !== FinancialEntityStatus.DISCONNECTED) {
          const lastFetchRecords = await this.lastFetchesPort.getByEntityId(
            entity.id,
          )
          lastFetch = Object.fromEntries(
            lastFetchRecords
              .filter(r => !!r.feature)
              .map(r => [r.feature as Feature, r.date]),
          )
        }
      } else {
        dictEntity.features = []
      }

      const entityVirtualImports = lastVirtualImportedEntities.get(entity.id)
      if (entityVirtualImports && entityVirtualImports.length > 0) {
        const virtualLastFetch = getLastFetchesForVirtual(entityVirtualImports)
        if (entity.origin === EntityOrigin.MANUAL) {
          lastFetch = virtualLastFetch
          dictEntity.features = Object.keys(virtualLastFetch) as Feature[]
        }
      }

      entities.push({
        ...dictEntity,
        status,
        connected: wallets,
        lastFetch,
        requiredExternalIntegrations:
          dictEntity.requiredExternalIntegrations ?? [],
        externalEntityId,
        virtualFeatures,
        nativelySupportedProducts: products ?? null,
      })
    }

    return { entities }
  }

  private async getLastVirtualImportsByEntity(): Promise<
    Map<string, VirtualDataImport[]>
  > {
    const lastVirtualImports =
      await this.virtualImportRegistry.getLastImportRecords()
    const lastVirtualImportedEntities = new Map<string, VirtualDataImport[]>()
    for (const virtualImport of lastVirtualImports) {
      if (!virtualImport.entityId) continue
      const list = lastVirtualImportedEntities.get(virtualImport.entityId) ?? []
      list.push(virtualImport)
      lastVirtualImportedEntities.set(virtualImport.entityId, list)
    }
    return lastVirtualImportedEntities
  }
}
